### PR TITLE
fix(cc): handle nloc==0 in DeepSpinPTExpt with phantom-atom padding

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
## Problem

Multi-rank spin MD can leave a rank with zero real local atoms (`nloc_real == 0`) when atoms migrate to other subdomains. The with-comm AOTI artifact hits an intermittent SIGFPE (integer divide by zero) at runtime in inductor-generated shape arithmetic that uses `nloc` as a divisor.

Reproduced on master CI run [`26667802665`](https://github.com/deepmodeling/deepmd-kit/actions/runs/26667802665):
```
Caught signal 8 (Floating point exception: integer divide by zero)
4  forward_lower_with_comm/.../wrapper.so(AOTInductorModel::run_impl+0xf482)
```

Root cause:
- The graph was traced with `nloc_min=1` (`serialization.py:362`) and inductor lowered an even stricter `nloc >= 2` runtime-check (visible in the generated `wrapper.cpp`'s `check_input_3`).
- That runtime-check is gated by env var `AOTI_RUNTIME_CHECK_INPUTS` (default OFF), so with `nloc = 0` the check is silently bypassed and the compiled graph runs through its own divide-by-zero on shape arithmetic.
- Whether the offending divide is actually emitted depends on inductor's code-gen choices, which vary across compiles — hence the intermittent nature.

## Fix

Prepend two phantom atoms with empty neighbour lists when `nloc_real == 0` so the AOTI graph runs with `nloc == 2` and never reaches the integer-divide-by-zero path. Phantoms have no neighbours so they contribute zero atomic energy / force / virial, preserving the physically-correct "this rank has no real atoms" result.

Key details (all in `source/api_cc/src/DeepSpinPTExpt.cc`):
- `dcoord` / `datype` / `dspin` get two zero-valued rows prepended.
- `firstneigh_tensor` gets two `-1` rows prepended (no neighbours).
- `mapping_tensor` gets two identity entries prepended.
- `comm_dict.nlocal` is set to `2` (not the LAMMPS-reported `0`) so `border_op` writes received ghost features past the phantom slots.
- Output arrays (`dforce`, `dforce_mag`, `datom_energy`, `datom_virial`) get the phantom prefix stripped before being scattered back to LAMMPS via `select_map`.

## Why phantoms rather than `Dim(min=0)` re-export

Bumping the trace constraint to `min=0` would require:
1. auditing every `nloc`-dependent divide in `deepmd/dpmodel/{descriptor,fitting,model}/` and protecting with `xp.maximum(nloc, 1)`;
2. `torch.export` re-emitting compatible guards (currently fails because spin-side shape relationships require `nloc >= 1` to be inferable);
3. inductor cooperating with the relaxed bound (it makes independent specialization choices downstream);
4. re-exporting every `.pt2` archive in `source/tests/infer/`.

The phantom approach is a strict superset of correctness and self-contained in one C++ file. The two approaches aren't mutually exclusive — the `min=0` route can land as a follow-up once the dpmodel audit is done.

## Test plan

- [x] Local CPU rebuild + `runUnitTests_cc --gtest_filter='*Spin*'`: **42 / 42 spin C++ regression tests pass** (12 TF-backend tests skipped, as expected in the PT-only venv).
- [ ] CI: the multi-rank LAMMPS test `test_pair_deepmd_mpi_dpa3_spin_empty_subdomain` should now pass deterministically. Local Python LAMMPS-MPI verification is blocked by a pre-existing OpenMPI/MPICH ABI mismatch in my local venv (the plugin's `ompi_mpi_*` symbols can't resolve against MPICH's `libmpi.so.12`), so end-to-end verification falls to CI.

## Known limitations

- The phantom path is structurally inert for `nloc_real > 0` (the `if (phantom_n > 0)` branch never fires), so the common path is unchanged.
- If a future inductor version bumps the `nloc` lower-bound to >2, `phantom_n` will need to track that minimum.
- This fix is in `DeepSpinPTExpt` only. The corresponding non-spin path in `DeepPotPTExpt` has the same code shape; non-spin DPA3 empty-subdomain currently passes in CI but could regress similarly with a future inductor change. Deferred to a follow-up if observed.
- Supersedes #5478 (which proposed skipping the test); this PR fixes the underlying bug instead.

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented crashes and incorrect results when some processes have no local atoms during distributed runs; placeholder (phantom) atoms are ignored in neighbor, force, energy, and per-atom outputs so reported values match real atoms.
* **Tests**
  * Updated tests to pass explicit per-atom parameters and exercise empty-subdomain and multi-rank behaviors.
* **Documentation**
  * Clarified test docstrings and model config comments about per-atom parameter handling.

### CODE REVIEW

File: source/api_cc/src/DeepSpinPTExpt.cc [hunk 1]
```diff
@@ -371,12 +371,43 @@ void DeepSpinPTExpt::compute(ENERGYVTYPE& ener,
...
+  //
+  // Fix: prepend two phantom atoms with no neighbours so the AOTI graph
+  // runs with ``nloc == 2``.  The phantoms have an empty nlist row and
+  // therefore contribute zero atomic energy / force / virial, preserving
+  // the physically-correct "this rank has no real atoms" semantics.
+  // ``nlocal`` in the comm tensors is set to ``2`` so border_op writes
+  // received ghost features past the phantom slots; outputs are stripped
+  // of the phantom prefix before being scattered back to LAMMPS atoms
+  // via ``select_map``.
+  const int phantom_n = (nloc_real == 0 && nall_real > 0) ? 2 : 0;
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc:393]: Addressed in commit `c43468fb7` (force-pushed). I went one step beyond a phantom-only subtract: a rank with no real local atoms contributes zero to the total energy by definition, so I just zero `ener` directly when `phantom_n > 0`.

The simpler-subtract approach would have been incomplete here — the spin path doubles atoms internally (real + spin half), so both halves carry the per-type-bias and zero-neighbour MLP output into `energy_redu`. `output_map["energy"]` only exposes the real half after the SpinModel's `[:, :nloc]` slice, so subtracting that alone would leave the spin half leaking through. (Confirmed empirically: an earlier attempt to subtract the real half showed mpi-2=-2.45 vs mpi-1=-1.49 in CI run 26796476553. The zero-out has no such residual.)

Forces / force_mag / virial don't need an analogous correction because phantom atomic outputs are coord-independent (no neighbours) so their derivatives are zero.

File: source/api_cc/src/DeepSpinPTExpt.cc [hunk 2]
```diff
@@ -445,11 +476,16 @@ void DeepSpinPTExpt::compute(ENERGYVTYPE& ener,
...
     if (lmp_list.mapping) {
       std::vector<std::int64_t> mapping(nall_real);
-      for (int ii = 0; ii < nall_real; ii++) {
-        mapping[ii] = fwd_map[lmp_list.mapping[bkw_map[ii]]];
+      for (int ii = 0; ii < phantom_n; ii++) {
+        mapping[ii] = ii;
+      }
+      for (int ii = phantom_n; ii < nall_real; ii++) {
+        mapping[ii] = fwd_map[lmp_list.mapping[bkw_map[ii - phantom_n]]];
       }
```
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc:506]: Not reachable in current code, so I'm not applying the suggestion. The reasoning:

1. The `if (lmp_list.mapping)` branch is **single-rank-only**. Both [`pair_deepmd.cpp:194`](https://github.com/deepmodeling/deepmd-kit/blob/c43468fb7477fe71a2db566e1c461454a658d011/source/lmp/pair_deepmd.cpp#L194) and [`pair_deepspin.cpp:204-257`](https://github.com/deepmodeling/deepmd-kit/blob/c43468fb7477fe71a2db566e1c461454a658d011/source/lmp/pair_deepspin.cpp#L204-L257) gate `lmp_list.set_mapping(...)` on `comm->nprocs == 1`. So in multi-rank, `lmp_list.mapping == nullptr` and we take the **else (identity-fallback)** branch instead.

2. `phantom_n > 0` is **multi-rank-only** — it only fires when `nloc_real == 0 && nall_real > 0`, which can only happen in a multi-rank LAMMPS run with an empty subdomain.

3. Combining: `phantom_n > 0` implies multi-rank, which implies `lmp_list.mapping == nullptr`, which means the branch you're flagging is never entered. The identity-fallback branch handles phantoms correctly because `mapping[ii] = ii` is self-consistent regardless of phantom offsets, and the with-comm AOTI artifact ignores `mapping` for ghost feature exchange anyway (`border_op` does that).

If a future change ever populates `lmp_list.mapping` in multi-rank, the fix would become applicable. Until then, applying it would be defensive against an unreachable scenario.

File: source/api_cc/src/DeepSpinPTExpt.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (wanghan-iapcm) [reply source/api_cc/src/DeepSpinPTExpt.cc:393]: Good catch — valid. Fixed in b025f9ac8: when `phantom_n > 0 && daparam > 0`, I now prepend `phantom_n * daparam` zero rows to `aparam_`, so the two phantom local atoms carry zero atomic parameters and the aparam tensor (`{1, nloc, daparam}`) stays aligned with the padded `nloc`. (`aparam_nall` is `false` on this exported spin path, so `aparam_` is a per-local-atom buffer.) Verified it compiles (`deepmd_cc`).

**Review by njzjz — APPROVED** [review]
